src/main/java/com/example/onlinebookstore/service/impl/OrderServiceImpl.java

```diff
@@ -0,0 +1,123 @@
+                .orElseThrow(() -> new EntityNotFoundException("Can't find user by id: " + userId));
+        ShoppingCart shoppingCart = shoppingCartRepository.findShoppingCartByUserId(userId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find shopping cart by this userId: " + userId));
+        Order order = createOrder(shoppingCart.getCartItems());
+
+        order.setUser(user);
+        order.setOrderDate(LocalDateTime.now());
+        order.setStatus(Order.Status.PENDING);
+        order.setShippingAddress(orderRequestDto.shippingAddress());
+        order.setTotal(calculateTotal(order.getOrderItems()));
+
+        shoppingCart.clearCartItems();
+        orderRepository.save(order);
+
+        return orderMapper.toResponseOrder(order);
+    }
+
+    @Transactional(readOnly = true)
+    @Override
+    public List<OrderResponseDto> getOrderHistory(Long id, Pageable pageable) {
+        Page<Order> orders = orderRepository.findAllByUserId(id, pageable);
+        return orders.stream()
+                .map(orderMapper::toResponseOrder)
+                .toList();
+    }
+
+    @Transactional(readOnly = true)
+    @Override
+    public List<OrderItemResponseDto> getOrderItems(Long orderId, Pageable pageable) {
+        Page<OrderItem> orderItems = orderItemRepository.findAllByOrderId(orderId, pageable);
+        return orderItems.stream()
+                .map(orderItemMapper::toResponseOrderItem)
+                .toList();
+    }
+
+    @Transactional(readOnly = true)
+    @Override
+    public OrderItemResponseDto getOrderItem(Long orderId, Long id) {
+        return orderItemMapper.toResponseOrderItem(orderItemRepository
+                .findByOrderIdAndId(orderId, id)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        String.format("Can't find order item by orderId: %d and itemId: %d",
+                                orderId, id))));
+    }
+
+    @Transactional
+    @Override
+    public OrderResponseDto updateOrderStatus(Long id, OrderStatusRequestDto requestDto) {
+        Order order = orderRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can't find order by this id: " + id));
+        order.setStatus(Order.Status.valueOf(requestDto.status()));
+        return orderMapper.toResponseOrder(order);
+    }
+
+    private Order createOrder(Set<CartItem> cartItemSet) {
+        Order order = new Order();
+        Set<OrderItem> orderItems = new HashSet<>();
+
+        for (CartItem cartItem : cartItemSet) {
+            OrderItem orderItem = new OrderItem();
+            orderItem.setQuantity(cartItem.getQuantity());
+            orderItem.setBook(cartItem.getBook());
+            orderItem.setPrice(cartItem.getBook().getPrice()
+                    .multiply(BigDecimal.valueOf(cartItem.getQuantity())));
+            orderItem.setOrder(order);
```

Comment:
this seem to be a job for mapper

you can even map a collection of cart items to a collection of order items